text/0328-polyglot-assert.md

```diff
@@ -0,0 +1,335 @@
+---
+rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/330
+tracking issue: https://github.com/aws/aws-cdk-rfcs/issues/328
+---
+
+# Polyglot Assert
+
+The `@aws-cdk/assert` module is only available to javascript users.
+
+The new polyglot assert module - `@aws-cdk/assertv2` - provides the same set of functionalities
+to users in all CDK supported languages.
+
+## Working Backwards
+
+### CHANGELOG
+
+**feat:** announcing assertv2: the assert library is now available in all CDK supported languages.
+
+### README
+
+<!--BEGIN STABILITY BANNER-->
+
+---
+
+![cdk-constructs: Stable](https://img.shields.io/badge/cdk--constructs-stable-success.svg?style=for-the-badge)
+
+---
+
+<!--END STABILITY BANNER-->
+
+> NOTE: This module contains *beta APIs*.
+>
+> All beta symbols are suffixed with the `Beta<n>`. When we have backwards
+> incompatible change, we will create a new symbol with a `Beta<n+1>` suffix
+> and deprecate the `Beta<n>` symbol.
+---
+This module allows asserting the contents of CloudFormation templates.
+
+To run assertions based on a CDK `Stack`, start off with -
+
+```ts
+const stack = new cdk.Stack(...)
+...
+const inspect = StackAssertionsBeta1.fromStackBeta1(stack);
+```
+
+Alternatively, assertions can be run on an existing CloudFormation template -
+
+```ts
+const file = '/path/to/file';
+const inspect = StackAssertionsBeta1.fromTemplateFileBeta1(file);
+```
+
+#### Full Template Match
+
+The simplest assertion would be to assert that the template matches a given
+template.
+
+```ts
+inspect.assertMatchTemplateBeta1({
+  Resources: {
+    Type: 'Foo::Bar',
+    Properties: {
+      Baz: 'Qux',
+    },
+  },
+});
+```
+
+#### Counting Resources
+
+This module allows asserting the number of resources of a specific type found
+in a template.
+
+```ts
+inspect.assertResourceCountBeta1('Foo::Bar', 2);
+```
+
+#### Resource Matching
+
+Beyond resource counting, the module also allows asserting that a resource with
+specific properties are present.
+
+The following code asserts that the `Properties` section of a resource of type
+`Foo::Bar` contains the specified properties -
+
+```ts
+inspect.assertResourceBeta1('Foo::Bar', {
+  Foo: 'Bar',
+  Baz: 5,
+  Qux: [ 'Waldo', 'Fred' ],
```

Comment:
What about complex matcher assertions? `objectLike`, `arrayWith` etc?

Reply:
The focus of this RFC was to discuss the basic premise of the APIs (and the beta suffix) and the design/implementation complexity.

It's in no way capturing all features of the assert that we will support. This will be added later.

Reply:
I would add `objectLike()` and `arrayWith()` as well at this point - it's interesting how to represent them in non-TypeScript languages.

Reply:
Not the focus of this RFC. We can discuss the ergonomics around this in a different avenue.